backend/app/usecase/search/search.go

```diff
@@ -66,13 +69,112 @@ func (s Search) searchResource(resource Resource, orderBy order.Order, query Que
 }
 
 func (s Search) searchShortLink(query Query, orderBy order.Order, filter Filter) (Result, error) {
-	return Result{}, nil
+	if query.User == nil {
+		return Result{}, errors.New("user not provided")
+	}
+	if len(query.Query) == 0 {
+		return Result{}, errors.New("query not provided")
+	}
+
+	shortLinks, err := s.getShortLinkByUser(*query.User)
+	if err != nil {
+		return Result{}, err
+	}
+
+	matchedShortLinks := matchShortLinks(shortLinks, query, orderBy)
+
+	filteredShortLinks := filterShortLinks(matchedShortLinks, filter)
+
+	return Result{
+		shortLinks: filteredShortLinks,
+		users:      nil,
+	}, nil
 }
 
 func (s Search) searchUser(query Query, orderBy order.Order, filter Filter) (Result, error) {
 	return Result{}, nil
 }
 
+func (s Search) getShortLinkByUser(user entity.User) ([]entity.ShortLink, error) {
+	aliases, err := s.userShortLinkRepo.FindAliasesByUser(user)
+	if err != nil {
+		return []entity.ShortLink{}, err
+	}
+
+	return s.shortLinkRepo.GetShortLinksByAliases(aliases)
+}
+
+func matchShortLinks(shortLinks []entity.ShortLink, query Query, orderBy order.Order) []entity.ShortLink {
+	var matchedAliasByAnd, matchedAliasByOr, matchedLongLinkByAnd, matchedLongLinkByOr []entity.ShortLink
+	keywords := strings.Split(query.Query, " ")
+	for _, shortLink := range shortLinks {
+		// check if query is contained in alias by "and" operator and by "or" operator
+		aliasByAnd, aliasByOr := stringContains(shortLink.Alias, keywords)
+		if aliasByAnd {
+			matchedAliasByAnd = append(matchedAliasByAnd, shortLink)
+			continue
+		} else if aliasByOr {
+			matchedAliasByOr = append(matchedAliasByOr, shortLink)
+			continue
+		}
+
+		// check if query is contained in long link by "and" operator and by "or" operator
+		longLinkByAnd, longLinkByOr := stringContains(shortLink.LongLink, keywords)
+		if longLinkByAnd {
+			matchedLongLinkByAnd = append(matchedLongLinkByAnd, shortLink)
+		} else if longLinkByOr {
+			matchedLongLinkByOr = append(matchedLongLinkByOr, shortLink)
+		}
+	}
+
+	// sort short links
+	sortShortLinks(orderBy, matchedAliasByAnd, matchedAliasByOr, matchedLongLinkByAnd, matchedLongLinkByOr)
+
+	// merge all the short links
+	return mergeShortLinks(matchedAliasByAnd, matchedAliasByOr, matchedLongLinkByAnd, matchedLongLinkByOr)
+}
+
+func filterShortLinks(shortLinks []entity.ShortLink, filter Filter) []entity.ShortLink {
+	if len(shortLinks) > filter.MaxResults {
+		shortLinks = shortLinks[:filter.MaxResults]
+	}
+
+	return shortLinks
+}
+
+func stringContains(s string, words []string) (bool, bool) {
+	and := true
+	or := false
+	for _, word := range words {
+		if !and && or {
+			return and, or
+		}
+
+		if strings.Contains(s, word) {
+			or = true
+		} else {
+			and = false
+		}
+	}
+	return and, or
+}
```

Comment:
This will probably need to be refactored. It returns two boolean results but it takes a slice of undetermined length which makes it a bit confusing. I'm having a bit of a hard time deciphering the and or comparisons. Could you provide a bit more context here please?

Reply:
I agree! What this function does is that it takes a string say "https://google.com/" and words ["google", "facebook"]. Then it checks if the string contains all the words, which here it doesn't and simultaneously if the string contains at least one word, which in the example it does contain the word google. 

My first thought was to optimize by traversing the words only once but I see now that this would be not a good practice. Therefore I decoupled now. Please check now